### ISSUE TITLE
Roslyn thinks that value type fields can be null if there's a `[MemberNotNull]` annotation

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#43751]
**Version Used**: 
```
commit aa691a174b9c809dd210fc43aea8982d0ee579d6
Merge: 995e2628339 81dbb931a2a
Author: Sam Harwell <sam.harwell@microsoft.com>
Date:   Mon Apr 27 23:19:58 2020 -0700

    Merge pull request #43715 from sharwell/intervaltree-any

    <F0><9F><90><87> Avoid using an ArrayBuilder for IntervalTree<T>.Any
```

**Steps to Reproduce**:

Compile the following code
```cs
using System.Diagnostics.CodeAnalysis;

class C
{
  public int i = 1;
  
  [MemberNotNull(nameof(i))] public void M() { }
}
```

**Expected Behavior**:
No warnings. `i` cannot contain a null value as it's `int`

**Actual Behavior**:
`CS8774: Member 'i' must have a non-null value when exiting.` is reported for `M`

**Notes**
In this particular example the annotation itself can have a warning but I believe that you can create a more reasonable example with generics e.g. something like this:
```
using System.Diagnostics.CodeAnalysis;
abstract class Base<T> where T: new()
{
  public T Prop { get; set; } = new T();
  [MemberNotNull(nameof(Prop))] public abstract string Init();
}

class Derived : Base<int?>
{
  public override string Init() => """";
}
```
Unfortunately I can't verift it now as `[MemberNotNull]` annotation is broken with generics e.g. https://github.com/dotnet/roslyn/issues/43749
